### DISCUSSION TITLE
Questions about version compatibility, migrations, etc.

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#880] [Q&A] [answered]
@ganigeorgiev I'm curious, are you using [semver](https://semver.org/) for versioning, or what kind of 'contract' are you planning for version numbering?

Major - ?
Minor - it looks like you are using this for breaking API changes and non-reversible migrations (eg, 0.7.9 to 0.8rc1)
Patch - bug fixes only - client SDK's and db schema will not change between patches, correct?

For pockethost.io, I'm trying to give users the ability to select which version of PocketBase to run. For example, if they are on 0.7.x, it would be nice to use the latest patch release unless they lock it in to a specific patch. But to move to 0.8, they would not be able to downgrade to 0.7 again because of the migrations. But I assume we would be safe auto-updating them to 0.8.x.

Along these same lines, when 0.8 is finalized, are you planning an auto-update or will users need to run the `pocketbase upgrade` command?

**Comment by ganigeorgiev** [comment, marked as answer]
@benallfree We use semantic versioning but nothing is stable yet before v1.0.0 so breaking changes (although I try to keep them minimal) could be expected between minor releases.

This is from the FAQ section in the above semver link:

> How should I deal with revisions in the 0.y.z initial development phase?
> The simplest thing to do is start your initial development release at 0.1.0 and **then increment the minor version for each subsequent release**.

Since we haven't reached v1.0.0 yet, I always apply the following rules for 0.X.Y:

- Increment Y for bugfixes only
- increment X for releases with new features (sometimes also combined with bugfixes)

> For pockethost.io, I'm trying to give users the ability to select which version of PocketBase to run.

It is better for new users to use directly PocketBase v0.8. For your use case, I would perform the upgrade by creating a copy of the old `pb_data` and create a new v0.8 instance with it. If later the user decide to downgrade for some reason, they could just switch to the old instance.

> Along these same lines, when 0.8 is finalized, are you planning an auto-update or will users need to run the pocketbase upgrade command?

Users will be expected to run the temporary upgrade command. This command will be removed (or repurposed) after v0.9.

**Reply by benallfree** [reply]
> It is better for new users to use directly PocketBase v0.8. 

My plan is that newly created instances will always use the latest version. 

Do you think it would be safe to allow the user to select up/down between `Y` patches? Those are backward and forward compatible, right? 

It seems unsafe to auto-upgrade them to the latest `X` version because of potential API incompatibility and inability to downgrade again if migrations are involved.

I'm thinking of introducing a `platform` concept, where the platform is `0.7` or `0.8` etc. Then I'll group available PocketBase versions to a platform and the user can select freely among those versions. This would be done just in case a certain patch made something worse and they wanted to switch back to a different version in the same platform.

Moving between platforms would be reversible only by restoring the backup. It would require a `pocketbase upgrade` command or similar, so they would be committed to the new platform unless they chose to restore their data to a backup point.

> If later the user decide to downgrade for some reason, they could just switch to the old instance.

Potentially with a loss of data, yes.

**Reply by ganigeorgiev** [reply]
> Do you think it would be safe to allow the user to select up/down between Y patches? Those are backward and forward compatible, right?

Most of the times yes, but I cannot promise full backward and forward compatibility before v1.0.0, because there may be cases where the fix for a bug may require breaking change (it's rare but it's possible).

**Reply by benallfree** [reply]
> there may be cases where the fix for a bug requires breaking change 

I guess it would probably make sense to bump `X` in those cases, but it won't matter much to me because I'll organize things by groups that are completely forward and backward compatible. That will give you the flexibility to version however is best, and I can give users the most stable hosted experience too.
